lxd/images.go

```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
 }
 
 func autoUpdateImages(ctx context.Context, d *Daemon) error {
-	projectNames := []string{}
+	imageMap := make(map[string][]db.Image)
+
 	err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
-		projects, err := tx.GetProjects(db.ProjectFilter{})
+		var err error
+
+		images, err := tx.GetImages(db.ImageFilter{AutoUpdate: true})
 		if err != nil {
 			return err
 		}
 
-		for _, project := range projects {
-			projectNames = append(projectNames, project.Name)
+		for _, image := range images {
+			imageMap[image.Fingerprint] = append(imageMap[image.Fingerprint], image)
 		}
 
 		return nil
 	})
 	if err != nil {
-		return errors.Wrap(err, "Unable to retrieve project names")
+		return errors.Wrap(err, "Unable to retrieve image fingerprints")
 	}
 
-	for _, project := range projectNames {
-		err := autoUpdateImagesInProject(ctx, d, project)
+	for fingerprint, images := range imageMap {
+		nodes, err := d.cluster.GetNodesWithImageAndAutoUpdate(fingerprint, true)
 		if err != nil {
-			return errors.Wrapf(err, "Unable to update images for project %s", project)
+			logger.Debug("Error getting nodes", log.Ctx{"fingerprint": fingerprint})
+			continue
+		}
+
+		var nodeIDs []int64
+
+		for _, node := range nodes {
+			err := d.cluster.Transaction(func(tx *db.ClusterTx) error {
+				var err error
+
+				nodeInfo, err := tx.GetNodeByAddress(node)
+				if err != nil {
+					return err
+				}
+
+				nodeIDs = append(nodeIDs, nodeInfo.ID)
+
+				return nil
+			})
+			if err != nil {
+				return errors.Wrap(err, "Unable to retrieve cluster node information")
+			}
+		}
+
+		// If multiple nodes have the image, select one to deal with it.
+		if len(nodeIDs) > 1 {
+			selectedNode, err := shared.GetStableRandomInt64FromList(len(images), nodeIDs)
+			if err != nil {
+				logger.Debug("Failed to get stable random value", log.Ctx{"err": err})
+				continue
+			}
+
+			if d.cluster.GetNodeID() != selectedNode {
+				continue
+			}
+		}
+
+		var newImage *api.Image
```

Comment:
@monstermunchkin as `newImage` isn't used outside of the `go func() {}` function it would make sense to reduce its scope and define it inside the function.

Also, my understanding of the go routine here is so that it allows for cancellation of the ongoing update and distribution operation in case the LXD daemon is stopped and the task is cancelled.

However just returning nil immediately doesn't clean up the go routine (and we can't know that the reason the task is being cancelled is because the LXD daemon is shutting down, the context could be cancelled for other reasons unbeknownst to this function). Also, if LXD daemon is shutting down (perhaps due to snap update) and the go routine is finished unexpectedly then we risk killing an unfinished image update or distribution operation unexpectedly. I believe this could cause some inconsistency that we would be aiming to fix with this PR (but I may be going beyond the scope of the PR that @stgraber intended so if so no worries).

But assuming cancelling the update and distribution process cleanly is in scope, then rather than immediately returning from this function pass the context to `autoUpdateImage` and `distributeImage` so they can finish up cleanly, and then return the error from `ctx.Err()` from those functions when context is cancelled, which you can then check here for being `context.Canceled` and if so return from the function.

I'm not sure how difficult it is to update `autoUpdateImage` and `distributeImage` to cleanly finish up on context cancellation, but it feels like adding the plumbing for it at least to those functions is worth doing now, even if they just have a comment saying "//TODO handling context cancellation properly" for now.

Reply:
@monstermunchkin I made the same comment on Tuesday, can you re-introduce the comment that this PR removed?

It used to say:
```
		// FIXME: since our APIs around image downloading don't support
		//        cancelling, we run the function in a different
		//        goroutine and simply abort when the context expires.
```